### DISCUSSION TITLE
코드의 출력 결과와 이유를 설명해주세요.

### DISCUSSION THREAD

**Discussion by pipisebastian** [publdaze/js-study#59] [스코프]
- 코드의 출력 결과와 이유를 설명해주세요.
```tsx
var x = 1;

function foo(){
	console.log(x);
	var x = 2;
}

foo();
console.log(x);
```

**Comment by pipisebastian** [comment]
> 코드의 출력 결과와 이유를 설명해주세요.

- `undefined`,`1` 
- 1번째 출력에서는 참조할 x를 찾기 위해, 해당 스코프를 검색합니다. 그런데 x의 경우 이미 함수가 실행되고, 호이스팅이 일어나 x는 undefined로 초기화가 되었기에 undefined를 출력합니다.
- 2번째 출력에서는 전역 스코프에서 검색하는데 , 1이 할당된 x를 찾았기에 1을 출력합니다.

**Comment by publdaze** [comment]
> - 코드의 출력 결과와 이유를 설명해주세요.
> ```tsx
> var x = 1;
> 
> function foo(){
> 	console.log(x);
> 	var x = 2;
> }
> 
> foo();
> console.log(x);
> ```

- foo 함수가 호출이 된 후 내부 코드가 실행되기 전 변수의 호이스팅이 일어남
- var 변수의 경우 선언 단계와 초기화 단계가 동시에 일어나고, 초기화 단계에서 쓰레기 값을 방지하기 위해 undefined가 할당 됨
- 따라서 첫번째 출력에서는 undefined가 출력됨
- 두 번째 출력의 경우 foo 함수의 스코프는 함수가 종료되고 더이상 참조하는 변수가 없기 때문에 스코프가 소멸되고, 전역 변수 x에 초기화 된 1이 최종적으로 출력 됨
